File: src/main/js/formulas/math/ROUND.ts
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class ROUND extends AbstractFormula {
  static SELF: ROUND = new ROUND();
  NAME = FormulaName.ROUND;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkLengthBetween(values.length, 1, 2, this.NAME);
    const value = Converters.first(this.collateralLookup(origin, values[0]));
    let placesRaw;
    if (values.length === 1) {
      placesRaw = 0.0;
    } else {
      placesRaw = Converters.toNumber(Converters.first(this.collateralLookup(origin, values[1])));
    }
    const n = Converters.toNumber(value);
    const places = placesRaw > 0 ? Math.floor(placesRaw) : Math.ceil(placesRaw);
    if (n < 0) {
      if (places < 0) {
        return Number(Math.round(parseFloat(n + "e-" + places * -1)) + "e" + places * -1);
      }
      return Number(Math.round(parseFloat(n + "e" + places)) + "e-" + places);
    }
    if (places < 0) {
      return Number(Math.round(parseFloat(n + "e-" + places * -1)) + "e" + places * -1);
    }
    return Number(Math.round(parseFloat(n + "e" + places)) + "e-" + places);
  }
}

File: src/main/java/io/protobase/f7/formulas/math/ROUND.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.function.BiFunction;
import java.util.function.Function;

public class ROUND extends AbstractFormula {
  public static FormulaName NAME = FormulaName.ROUND;
  public static ROUND SELF = new ROUND();

  public ROUND() {
    super();
  }

  public ROUND(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkLengthBetween(values.length, 1, 2, NAME);
    Object value = Converters.first(collateralLookup.apply(origin, values[0]));
    if (values.length == 1) {
      return ((double) Math.round(Converters.toDouble(value)));
    }
    Double placesRaw = Converters.toDouble(Converters.first(collateralLookup.apply(origin, values[1])));
    int places;
    if (placesRaw > 0) {
      places = new Double(Math.floor(placesRaw.floatValue())).intValue();
    } else {
      places = new Double(Math.ceil(placesRaw.floatValue())).intValue();
    }
    BigDecimal bd = new BigDecimal(Converters.toDouble(value));
    bd = bd.setScale(places, RoundingMode.HALF_UP);
    return bd.doubleValue();
  }
}
